source/common/http/http1/balsa_parser.cc

```diff
@@ -13,9 +13,67 @@ namespace Envoy {
 namespace Http {
 namespace Http1 {
 
+namespace {
+
 using ::quiche::BalsaFrameEnums;
 using ::quiche::BalsaHeaders;
 
+constexpr absl::string_view kColonSlashSlash = "://";
+
+// This method is crafted to match the URL validation behavior of the http-parser library.
+bool isUrlValid(absl::string_view url, bool is_connect) {
+  if (url.empty()) {
+    return false;
+  }
+
+  // Same set of characters are allowed for path and query.
+  const auto is_valid_path_query_char = [](char c) {
+    return c == 9 || c == 12 || ('!' <= c && c <= 126);
```

Comment:
This is to match http-parser behavior as closely as possible so that there is little friction in changing to Balsa.  After the transition is complete, we can start aligning validation code with the spec, I imagine that would be part of the unified header validation project.

http-parser has this at https://github.com/nodejs/http-parser/blob/main/http_parser.c#L247, which is a weird bitmap of allowed characters.